Updates to better support ormsgpack

timfel opened the pull request:
@msimacek These changes make the diff for supporting ormspack significantly smaller, can you check if you see any issues proposing this upstream?

This will also reduce the orjson diff a good chunk

davidhewitt committed:
ci: reduce number of jobs & improve coverage measurement (#5052)

* ci: run coverage as part of "build" job

* merge free-threaded tests into main build jobs

* make `test-introspection` jobs only run on PRs if requested

* fix 3.7 & 3.8 build error

* fix `check-feature-powerset`

timfel committed:
Expose the vectorcall functions also on GraalPy

msimacek commented on pyo3-ffi/src/cpython/listobject.rs:
```diff
@@ -28,9 +28,18 @@ pub unsafe fn PyList_GET_ITEM(op: *mut PyObject, i: Py_ssize_t) -> *mut PyObject
+    #[cfg(GraalPy)]
+    {
+        let previous = crate::PyList_GetItem(op, i);
+        Py_XDECREF(previous);
```
This seems wrong. The whole point of `PyList_SET_ITEM` is that it doesn't decref the existing items because they are not initialized. Here you even decref it twice, once manually, second time in `PyList_SetItem`. What was wrong with using `_PyList_SET_ITEM`?

timfel replied:
Right, I confused myself

msimacek commented on pyo3-ffi/src/cpython/tupleobject.rs:
```diff
@@ -21,16 +23,26 @@ pub unsafe fn PyTuple_GET_SIZE(op: *mut PyObject) -> Py_ssize_t {
+    #[cfg(any(GraalPy, PyPy))]
+    {
+        let previous = PyTuple_GetItem(op, i);
+        Py_XDECREF(previous);
```
Ditto for the two decrefs where there shouldn't be any.

timfel replied:
Confused myself, changed to Py_XINCREF to counter the decref in `PyTuple_SetItem`

msimacek replied:
The memory might be unitialized, the decref in `PyTuple_SetItem` could segfault. It's unfortunate that we didn't leave any good symbol to use for this. We only have `PyTruffleTuple_GetItems`, I'm not sure we want to expose that.

timfel replied:
It shouldn't be uninitialized though. If it's managed, we used `calloc` to get it to native, and the native allocator does `memset(obj, '\0', size);`

msimacek replied:
If it's a native subclass it can use its own allocator that doesn't clear the memory. And even when not, the incref afterwards is just a terrible workaround and reliance on implementation details. Could you export a proper function for this on GraalPy side and make this graalpy 25+?

Owen-CH-Leung committed:
Impl time crate into pyo3 (#5057)

* First draft for implementing time crate into PyO3

* Refactor FromPyObject for PrimitiveDateTime. Disable proptest

* Refactor FromPyObject for OffsetDateTime and UtcDateTime

* Create extract_date_time function for reusable code

* Fix all proptests

* Fix doc test

* Add changelog

* Downgrade time crate version for msrv. Fix clippy error

* Remove UtcDateTime

* Fix fmt CI failure

* Bring back UtcDateTime. Bump time crate to 0.3.38. Add features doc.

* Add time feature to noxfile

* use timezone_utc. add const SECONDS_PER_DAY. Revise doc example

* Fix FromPyObject for UtcDateTime

* `intern` only used on limited api

---------

Co-authored-by: Icxolu <10486322+Icxolu@users.noreply.github.com>

dependabot[bot] committed:
Bump codecov/codecov-action from 4 to 5 (#5072)

Bumps [codecov/codecov-action](https://github.com/codecov/codecov-action) from 4 to 5.
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/v4...v5)

---
updated-dependencies:
- dependency-name: codecov/codecov-action
  dependency-version: '5'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

davidhewitt committed:
fix free-threading issue in `test_coroutine` (#5069)

Tpt committed:
Introspection: add function signatures (#5025)

* Introspection: add function signatures

No annotations or explicit default values yet

Fixes an issue related to object identifiers path

* Better default value

* Refine arguments struct

* Introduce VariableLengthArgument

* Adds pyfunctions tests

* Adds some serialization tests

davidhewitt committed:
release notes for 0.24.1 and 0.24.2 (#5083)

* release: 0.24.1

* release: 0.24.2

exg committed:
ffi: expose PyDateTime_*_GET_TZINFO on PyPy (#5079)

They are available in PyPy3.10 and PyPy3.11.

bschoenmaeckers committed:
add `PyModule_Add` to `pyo3-ffi` (#5085)

* add `PyModule_Add` to `pyo3-ffi`

* add `PyModule_AddObjectRef` to `pyo3-ffi`

davidhewitt committed:
remove private types from `pyo3-ffi` (#5064)

* remove private types from `pyo3-ffi`

* newsfragments

* make `PyCodeObject` opaque on all versions

* fix msrv

* fix msrv, try 2

* remove unused constant

* sync `cpython/compile.rs` with header at 3.13

* fix size of INT_MAX

* remove `PyCode_GetNumFree`

prutschman-iv committed:
docs: Include example of calling with PyDict for kwargs (#5078)

* Include example of calling with PyDict for kwargs

Resolve PyO3/pyo3#1462

* remove empty line

---------

Co-authored-by: Icxolu <10486322+Icxolu@users.noreply.github.com>

Icxolu committed:
remove `AsPyPointer` trait (#5071)

* remove `AsPyPointer` trait

* fix typo

Co-authored-by: Bruno Kolenbrander <59372212+mejrs@users.noreply.github.com>

---------

Co-authored-by: Bruno Kolenbrander <59372212+mejrs@users.noreply.github.com>

bschoenmaeckers committed:
Add `PyTzInfo` constructors (#5055)

* Add `PyTzInfo` constructors

* Make `PyTzInfo::utc` failable

* Deprecate `timezone_utc`

davidhewitt committed:
ci: fix clippy jobs only running Python 3.7 checks (#5090)

* ci: fix clippy jobs only running Python 3.7 checks

* pass correct config file for 3.13t

Tpt committed:
Fixes pytests tests when using Python 3.7 (#5095)

* Disable positional only test in Python 3.7

* Update pytests/tests/test_pyfunctions.py

Co-authored-by: Nathan Goldbaum <nathan.goldbaum@gmail.com>

* fix ruff

---------

Co-authored-by: Nathan Goldbaum <nathan.goldbaum@gmail.com>

Icxolu committed:
remove deprecated `from_py_with` string variant (#5097)

0x676e67 committed:
docs: fix typo in `Ipv4Addr` and `Ipv6Addr` type names (#5098)

clin1234 committed:
Bump supported cpython version to 3.14 for testing (#4811)

* Add news item

* Move news file

* Fix version limit check in noxfile.py

* Bump Python version for testing debug builds

* 3.14 is available from GH's setup-python action

* Bump maximum supported CPython version in pyo3-ffi

* Rework PyASCIIObject and PyUnicodeObject to be compatible with 3.14

Due to python/cpython#128196, data types within `PyASCIIObject.state` have changed, resulting in test failures when building against 3.14.

* Run `cargo fmt --all`

* Actually add Py_3_14 as a legitimate macro

When `rustc` is invoked, the macro is included with the `--check-cfg`
flag, but not with the `--cfg` flag. This caused errors about duplicate
definitions to spew out when building with stable Rust toolchains.

* Revert "Actually add Py_3_14 as a legitimate macro"

This reverts commit 5da57afaee1975643bcf8e61dc9b1ce9d320d47c.

* Fix version macro placement for 3.14-specific getters and setters

* Import 'c_ushort' only if compiling against CPython 3.14 or later

* Add wrapper functions for the statically_allocated field

* Remove unused libc::c_ushort

* Add (hopefully) final version-specific macros

* Port 3.14-specific 64-bit code of Py_INCREF

* Don't expose PyDictObject.ma_version_tag when building against 3.14 or later

* fix ffi-check on the GIL-enabled ABI

* fix older pythons

* fix ffi-check on older pythons

* WIP: update for 3.14t

* fix ffi-check on the free-threaded build

* fix clippy

* fix clippy on older python versions

* fix cargo check on the MSRV

* fix ffi-check on 3.13t

* fix CI which is using 3.13.1

* fix copy/paste error in noxfile

* update ffi bindings for the latest changes in 3.14

* update layout of refcnt field on gil-enabled build

* delete unused HangThread struct

* fix ffi-check on GIL-enabled build

* Revert "delete unused HangThread struct"

This reverts commit 3dd439d90efc05ad7a678b9b81f0348a7d3f921f.

* config-out HangThread

* fix 3.13 ffi-check

* fix debug python build error

* fix graalpy build

* Ignore DeprecationWarnings from the  pytest_asyncio module in tests

* Add abi3-py314

* fix free-threading issue in `test_coroutine` (#5069)

* Introspection: add function signatures (#5025)

* Introspection: add function signatures

No annotations or explicit default values yet

Fixes an issue related to object identifiers path

* Better default value

* Refine arguments struct

* Introduce VariableLengthArgument

* Adds pyfunctions tests

* Adds some serialization tests

* respond to david's code review

* add comment and fix test failure

* fix check-feature-powerset

* fix clippy

* fix wasip1 clippy

* fix 32 bit python 3.14 bug

* mark test-py step continue-on-error for dev python builds

* use github issue URL

* run ffi-check before running tests

* fix ffi-check for 3.14.0a7

---------

Co-authored-by: Nathan Goldbaum <nathan.goldbaum@gmail.com>
Co-authored-by: David Hewitt <mail@davidhewitt.dev>
Co-authored-by: Thomas Tanon <thomas.pellissier-tanon@helsing.ai>

bschoenmaeckers committed:
add `Py_HashBuffer` to `pyo3-ffi` (#5086)

* add `Py_HashBuffer` to `pyo3-ffi`

* Add `Py_HashBuffer` binding when 3.14+

* Move to `cpython/pyhash.rs`

* Update changelog

* Enable`Py_HashBuffer` compat on 3.14 abi3

bschoenmaeckers committed:
remove TODO in `pyo3-build-config` (#5102)

Icxolu committed:
add declarative module level constants export (#5096)

davidhewitt committed:
ci: tidy up some `with_gil(|gil|` -> `with_gil(|py|` (#5104)

Icxolu committed:
introduce `immutable_type` option for pyclasses (#5101)

* introduce `immutable_type` option for pyclasses

* allow `immutable_type` from 3.10 on the unlimited api

* fix `is_abi3_before` check

---------

Co-authored-by: David Hewitt <mail@davidhewitt.dev>

davidhewitt committed:
remove FFI definitions of private static variables (#5105)

* remove FFI definitions of private static variables

* newsfragment

* revert removal of `_PyWeakRef_RefType` for now

davidhewitt committed:
ci: accept 429 as valid response for external url (#5106)

* ci: accept 429 as valid response for external url

* fixup syntax

m-ou-se committed:
Add #[track_caller] to with_gil. (#5109)

* Add #[track_caller] to with_gil.

* Add newsfragment.

m-ou-se committed:
Remove cfg-if dependency. (#5110)

* Remove cfg-if dependency.

It was only used once.

* Add newsfragment.

Icxolu committed:
unify `IntoPyObject`/`FromPyObject` derive attributes (#5070)

m-ou-se committed:
Use thread::park() in HangThread. (#5115)

* Use thread::park() in HangThread.

std::thread::park() is basically Rust's version of libc::pause(), it
waits for a 'signal' from std::thread::unpark().
On its own it can be used as a way to block forever without needing to
resort to sleep(9999999) or platform-specific apis.

* Add newsfragment.

davidhewitt committed:
bump supported GraalPy version to 24.2 (#5116)

* bump supported GraalPy version to 24.2

* newsfragment

clin1234 committed:
docs: Document how to add support CPython versions (#5113)

* docs: Document how to add support CPython versions

* Update Contributing.md

Co-authored-by: David Hewitt <mail@davidhewitt.dev>

* Update Contributing.md

Co-authored-by: David Hewitt <mail@davidhewitt.dev>

* Update Contributing.md

Co-authored-by: David Hewitt <mail@davidhewitt.dev>

* docs: Update Contributing.md

Co-authored-by: Icxolu <10486322+Icxolu@users.noreply.github.com>

---------

Co-authored-by: David Hewitt <mail@davidhewitt.dev>
Co-authored-by: Icxolu <10486322+Icxolu@users.noreply.github.com>

timfel committed:
Call Py_Is function on GraalPy

clin1234 committed:
Beta fix (#5123)

* Uncomment initconfig.rs for 3.14 beta1

* Uncomment pyerrors.rs for 3.14 beta 1

* Create 5123.fixed.md

bschoenmaeckers committed:
fix `jiff::Zoned` into python conversion when `zoneinfo` is not available (#5120)

* fix `jiff::Zoned` `IntoPyObject` when `zoneinfo` is not available

* Fallback on backports.zoneinfo for python <3.9 when converting timezones

clin1234 committed:
docs: Fix md formatting on checkboxes (#5122)

davidhewitt committed:
ci: move main build to use `test-rust` nox job (#5091)

* ci: move main build into nox

* remove `MAYBE_NIGHTLY` flag

davidhewitt committed:
tidy up migration.md ahead of 0.25 release (#5129)

davidhewitt committed:
revert tz change for jiff test (#5130)

superserious-dev committed:
Implement optional feature `ordered-float` for NotNan/OrderedFloat <-> Python float conversions (#5114)

* Implement OrderedFloat & NotNan conversions for f32 & f64

* Reduce test boilerplate with macros for the wrapper types

* Make roundtripped zero test stricter on both Rust & Python side

* Simplify conversion implementations using macros

* Add documentation for feature

* Add newsfragment

* Remove unneeded closure

* Fix typo

* Remove ffi call

* Use exact version in Cargo.toml instead of lower bound

* Use `py_run!`, `py.eval`, and `py.import` to simplify code

* Add type annotations for WASM due to `py_run!`

* Change newsfragment filename to use Pull Request ID instead of Issue ID

* Add missing #[test] annotation for wasm test

bschoenmaeckers committed:
Add `PyRange` wrapper (#5117)

* Add `PyRange` wrapper with conversions to/from `std::ops::Range`

* `IntoPyObject` for `RangeFull` `RangeFrom` `RangeTo` `RangeToInclusive`

* Remove conversions & address comments

clin1234 committed:
Fix clippy errors (#5134)

* Fix clippy errors

* Add news file

* Delete newsfragments/5134.fixed.md

IvanIsCoding committed:
Add `#[pyclass(generic)]`  (#4926)

* First attempt at supporting pyclass(generic)

* Include class_geitem in macro

* Minor formatting

* Pass py argument to new()

* Add tests covering pyclass(generic)

* Add changelog entry

* Add test covering __getitem__

* Fix compilation error tests

* Fix Alias -> TypingAlias

* Make test case slightly stronger

* Add compilation error test

* Do not allow pyclass(generic) for enums

* Remove unused import

* Add import only to 3.9+

* Add documentation

* Ignore Python typing hints newly added Rust code

* Fix clippy warnings for the test

* Add newline

* Update TRYBUILD tests

* Move `generic` in table

* TypingAlias is only available in 3.10+

* Update trybuild test

davidhewitt committed:
more format args inlining (#5135)

* more format args inlining

* cargo fmt

* even more inlining

* also fix time feature, more coverage

ngoldbaum committed:
Don't expose PyASCIIObjectState on Python3.14 and newer (#5133)

* fix FFI PyUnicode bindings for 3.14.0b1

* Remove PyASCIIObjectState from pyo3-ffi for Py_3_14

* Update pyo3-ffi/src/cpython/unicodeobject.rs

Co-authored-by: David Hewitt <mail@davidhewitt.dev>

---------

Co-authored-by: David Hewitt <mail@davidhewitt.dev>

bschoenmaeckers committed:
Implement `OnceExt` & `MutexExt` for `parking_lot` & `lock_api` (#5044)

* Implement `OnceExt` & `MutexExt` for `parking_lot` & `lock_api`

* Implement `MutexExt` for `Arc<lock_api::Mutex>`

* Move `MutexExt` lock result into type parameter

* Remove lifetime from associated type of `OnceExt`

* Add docs

* add changelog

* Add new features to `full`

* mention `parking_lot` in guide

* Add tests

* Update docs to mention all cargo features

Co-authored-by: Nathan Goldbaum <nathan.goldbaum@gmail.com>

* Also test `parking_lot::Mutex::lock_py_attached` on `lock_api` feature

* Update newsfragments/5044.added.md

Co-authored-by: Nathan Goldbaum <nathan.goldbaum@gmail.com>

* Add additional case to `once_ext` test

* enable `arc_lock` feature on `parking_lot` during testing

---------

Co-authored-by: Nathan Goldbaum <nathan.goldbaum@gmail.com>

Icxolu committed:
add `rename_all` support for `#[derive(IntoPyObject)]` (#5112)

davidhewitt committed:
ci: reduce verbosity & install zoneinfo backport on <3.9 (#5136)

* ci: fixup verbosity & install zoneinfo backport on <3.9

* correct typo

* try different syntax for `verbose`

Tpt committed:
Implements basic method introspection (#5087)

* Implements basic method introspection

* Code review feedback

* Fixes CI

davidhewitt committed:
release: 0.25.0 (#5128)

davidhewitt committed:
docs: fixup typo on object.md (#5143)

timfel committed:
Add PyBytes_AS_STRING, seen in ormsgpack

Icxolu committed:
ci: updates for Rust 1.87 (#5147)
